File: javascript/observer/WeatherRama/subject/WeatherStation.js
import {Subject} from "./Subject.js";
import {Observer} from "../observer/Observer.js";

export class WeatherStation extends Subject {

    // private variables
    #observers
    #temperature
    #humidity
    #pressure

    constructor() {
        super();
        this.#observers = []
    }

    registerObserver(newObserver) {
        super.registerObserver(newObserver);
        this.#observers.push(newObserver)
    }

    removeObserver(removeObserver) {
        super.removeObserver(removeObserver);
        let index = this.#observers.indexOf(removeObserver)
        this.#observers.splice(index, 1)
    }

    notifyObservers() {
        super.notifyObservers();
        let observer = new Observer()
        for (observer of this.#observers) {
            observer.update(this.#temperature, this.#humidity, this.#pressure)
        }
    }

    setMeasurements(temperature, humidity, pressure) {
        this.#temperature = temperature
        this.#humidity = humidity
        this.#pressure = pressure
        this.notifyObservers()
    }

}

File: ruby/observer/WeatherRama/subject/weather_station.rb
require './observer/observer'
require './subject/subject'

class WeatherStation < Subject

  def initialize
    super
    @observers = []
    @temperature = 0
    @humidity = 0
    @pressure = 0
    @observer = Observer
  end

  def register_observer(new_observer)
    super
    @observers.push(new_observer)
  end

  def remove_observer(remove_observer)
    super
    @observers.delete(remove_observer)
  end

  def notify_observers
    super
    for @observer in @observers
      @observer.update(@temperature, @humidity, @pressure)
    end
  end

  def set_measurements(temperature, humidity, pressure)
    @temperature = temperature
    @humidity = humidity
    @pressure = pressure
    notify_observers
  end

end